keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,510 @@
+
+Hence, we need a mechanism to express that in the OpenAPI validation schema (compare https://github.com/kubernetes/kubernetes/pull/64558#issuecomment-403564033).
+
+**Raw JSON Option 1:** add a format `json`, e.g.:
+
+```json
+{"properties": {"x": {"format": "json"}}}
+```
+
+In this example "x" is excluded from pruning. Note that you can still use any kind of OpenAPI validation schema constructs to restrict `"x"` further.
+
+Note that we lose expressiveness of the existing `format` strings: we either apply the format to `"json"` or any other pre-defined format. I.e. we cannot express that pruning should be disabled, but if it is an integer, it should be an int32.
+
+**Question:** this feels like a reasonable loss of expressivity. Do we accept that?
+
+**Raw JSON Option 2:** add an extension property, e.g.
+
+```json
+{"properties": {"x": {"x-no-pruning": true}}}
+```
+
+This would not lead to a loss in expressivity. It is formulated negatively intentionally to have `false` as the default with `omitempty`.
+
+As OpenAPI extensions are prefixed with `x-kubernetes` elsewhere, we probably should name it `x-kubernetes-no-pruning: true`.
+
+**We propose to follow option with the `x-kubernetes` prefix** because it is more explicit, does not reduce expressivity and does not mangle with the already very vague `format` field definition.
+
+### Nested x-no-pruning
+
+**Question:** should we support nested `x-no-pruning`, i.e. disabling pruning for a sub-object, but re-enable it something deep inside of it? E.g.
+
+```json
+{
+    "properties": {
+        "x": {
+            "x-no-pruning": true,
+            "properties": {
+		        "y": { 
+                    "x-kubernetes-no-pruning": false,
+                    "properties": { "z": {} }
+                }
+            }
+        }
+    }
+}
+```
+
+If we do, the object 
+
+```json
+{
+    "a": 1,
+    "x": {
+        "b": 2,
+        "y": {
+            "c": 3,
+            "z": 42
+        }
+    }
+}
+```
+
+is pruned to  `{"x":{"b": 2, "y":{"z": 42}}}`.
+
+## Opt-in and Opt-out of Pruning on CRD Level
+
+We will add a pruning flag to `CustomResourceDefinitionSpec` of `apiextensions.k8s.io/v1beta1`:
+
+```golang
+type CustomResourceDefinitionSpec struct {
+  ...
+	
+  // Prune enables pruning of unspecified fields. Defaults to false.
+  // Note: this will default to true in version v1.
+  Prune *bool
+}
+```
+
+I.e. for `apiextensions.k8s.io/v1beta1` this will default to `false`.
+
+For `apiextensions.k8s.io/v1` we will change the default to `true` and forbid `false` during creation and updates. In `v1` the only way to opt-out from pruning is via setting `x-kubernetes-no-prune: true` in the schema.
```

Comment:
Are you going to automatically propagate this with a version conversion function? Otherwise I don't see how you avoid a sudden behavior shift.